src/PhpDocumentFactory.php

```diff
@@ -0,0 +1,215 @@
+use LanguageServer\Protocol\DiagnosticSeverity;
+use phpDocumentor\Reflection\DocBlockFactory;
+use PhpParser\ErrorHandler\Collecting;
+use PhpParser\NodeTraverser;
+use PhpParser\NodeVisitor\NameResolver;
+use Sabre\Event\Promise;
+use function Sabre\Event\coroutine;
+
+class PhpDocumentFactory
+{
+    /**
+     * Holds account of all definitions, references etc.
+     * @var Project
+     */
+    private $project;
+    
+    /**
+     * Instance of the PHP parser
+     *
+     * @var Parser
+     */
+    private $parser;
+
+    /**
+     * The DocBlockFactory instance to parse docblocks
+     *
+     * @var DocBlockFactory
+     */
+    private $docBlockFactory;
+
+    /**
+     * The DefinitionResolver instance to resolve reference nodes to Definitions
+     *
+     * @var DefinitionResolver
+     */
+    private $definitionResolver;
+
+    /**
+     * Reference to the language server client interface
+     *
+     * @var LanguageClient
+     */
+    private $client;
+
+    /**
+     * The client's capabilities
+     *
+     * @var ClientCapabilities
+     */
+    private $clientCapabilities;
+
+    public function __construct(LanguageClient $client, Project $project, ClientCapabilities $clientCapabilities = null)
+    {
+        $this->client = $client;
+        $this->clientCapabilities = $clientCapabilities ? : new ClientCapabilities();
+        $this->parser = new Parser;
+        $this->docBlockFactory = DocBlockFactory::createInstance();
+        $this->definitionResolver = new DefinitionResolver($project);
+        $this->project = $project;
+    }
+
+    /**
+     * Loads the document by doing a textDocument/xcontent request to the client.
+     * If the client does not support textDocument/xcontent, tries to read the file from the file system.
+     * The document is NOT added to the list of open documents, but definitions are registered.
+     *
+     * @param string $uri
+     * @return Promise <PhpDocument>
+     */
+    public function loadDocument(string $uri): Promise
+    {
+        return coroutine(function () use ($uri) {
+            $limit = 150000;
+            if ($this->clientCapabilities->xcontentProvider) {
+                $content = (yield $this->client->textDocument->xcontent(new Protocol\TextDocumentIdentifier($uri)))->text;
+                $size = strlen($content);
+                if ($size > $limit) {
+                    throw new ContentTooLargeException($uri, $size, $limit);
+                }
+            } else {
+                $path = uriToPath($uri);
+                $size = filesize($path);
+                if ($size > $limit) {
+                    throw new ContentTooLargeException($uri, $size, $limit);
+                }
+                $content = file_get_contents($path);
+            }
+            
+            return $this->createDocument($uri, $content, false);
+        });
+    }
+
+    /**
+     * Ensures a document is loaded and added to the list of open documents.
```

Comment:
I find it a bit weird that a `createDocument` method of a factory causes side-effects. I would assume it to only create a document.